### ISSUE TITLE
Some clients aren't lowercasing hashtags

### ISSUE THREAD

**Issue by alexgleason** [nostrability/nostrability#55]
## client initiating action

I don't know

## client receiving or observing action (or lack thereof)

Every. As an example, try:

```
nak req -t t=Bitcoin relay.primal.net
```

## what happens 

Trending algorithms and hashtag feeds do not work properly because of these stupid clients. Since Nostr events are immutable I can't just patch over this problem, and I can't make all tag queries case insensitive in my relay either.

**Comment by alexgleason** [comment]
I am going to just start blocking all events that include uppercase "t" tags on my relays. I am already blocking any events with "invisible" hashtags, eg they include a `t` tag but no matching `#hashtag` in the text. The majority of this is spam.

**Comment by alltheseas** [comment]
For reference Twitter changes all to proper case. For instance:

#OLYMPICS -> #Olympics
#olympics -> #Olympics
#olymPics -> #Olympics

**Comment by alexgleason** [comment]
@alltheseas The case in the text can be any, but there's also a `t` tag inserted into the tags that you can query by in filters. And that has to be lowercase, otherwise you'd have to query every permutation of the word to get all posts.

**Comment by alltheseas** [comment]
10-4 thanks for the lesson ser :pray:

**Comment by alltheseas** [comment]
Changed title and added label to mention t-tag

**Comment by jb55** [comment]
yeah this is one of those things which makes a lot of sense but isn't
explicitly mentioned anywhere in the NIPs

**Comment by alltheseas** [comment]
@fiatjaf where is the best place for this? NIPs? elsewhere ?

**Comment by AsaiToshiya** [comment]
Is this a different issue from https://github.com/nostr-protocol/nips/commit/d0962aac09ed1747c19bb96ce4accdaaab707c8d?

**Comment by alexgleason** [comment]
> Is this a different issue from [nostr-protocol/nips@d0962aa](https://github.com/nostr-protocol/nips/commit/d0962aac09ed1747c19bb96ce4accdaaab707c8d)?

Same issue. That fixes it not being in the NIPs. Still don't know which clients are doing it though.

**Comment by alltheseas** [comment]
We can keep in here as "recommend practice" for now, unless fiatjaf or NIP devs merge in the spec

**Comment by fabianfabian** [comment]
FYI Nostur right now adds the tag the way the author wrote it and then also always includes a lowercased version.

**Comment by jb55** [comment]
On Wed, Aug 21, 2024 at 04:17:13AM -0700, Fabian wrote:
>FYI Nostur right now adds the tag the way the author wrote it and then also always includes a lowercased version.

yes this is the correct way to do it

**Comment by alltheseas** [comment]
added best-practice label

**Comment by alexgleason** [comment]
> FYI Nostur right now adds the tag the way the author wrote it and then also always includes a lowercased version.

This actually makes sense, so I might be wrong.

In my original post I was complaining about events which only include the uppercase version and no lowercase version. A lowercase version must be included. But I think it's reasonable to also include the original version.

**Comment by jb55** [comment]
Alex Gleason ***@***.***> writes:

>> FYI Nostur right now adds the tag the way the author wrote it and then also always includes a lowercased version.
>
> This actually makes sense, so I might be wrong.
>
> In my original post I was complaining about events which only include the uppercase version and no lowercase version. A lowercase version must be included. But I think it's reasonable to also include the original version.

why?

**Comment by alltheseas** [comment]
@mmalmi reports that NIP-99 marketplace #65 app(s) are not lowercasing t-tags
